docs/proposals/20190610-machine-states-preboot-bootstrapping.md

```diff
@@ -0,0 +1,446 @@
+MachinePhaseProvisioned = MachinePhaseType("provisioned")
+```
+
+#### Transition Conditions
+- `Machine.Status.Phase` is set to “Provisioning”
+- `Machine.Spec.InfrastructureRef`->`Status.Phase` is “Ready”
+
+#### Expectations
+- Machine’s infrastructure has been created and the compute resource is available to be configured.
+
+---
+### Running
+```go
+// MachinePhaseRunning is the Machine state when it has
+// become a Kubernetes Node in a Ready state.
+MachinePhaseRunning = MachinePhaseType("running")
+```
+
+#### Transition Conditions
+- `Machine.Status.Phase` is set to “Provisioned”
+- `Machine.Spec.ProviderID` is not `<nil>`
+- A Kubernetes Node is found with the same `Machine.Spec.ProviderID` and its state is “Ready”
+
+#### Expectations
+- Machine controller should set `Machine.Status.NodeRef`.
+
+---
+### Deleting
+```go
+// MachinePhaseDeleting is the Machine state when a delete
+// request has been sent to the API Server,
+// but its infrastructure has not yet been fully deleted.
+MachinePhaseDeleting = MachinePhaseType("deleting")
+```
+
+#### Transition Conditions
+- `Machine.Phase` can be any value
+- `Machine.ObjectMeta.DeletionTimestamp` is not `<nil>`
+
+#### Expectations
+- Node associated should be drained and cordoned.
+- Machine’s related resources should be deleted first using cascading deletion.
+
+---
+### Deleted
+```go
+// MachinePhaseDeleted is the Machine state when the object
+// and the related infrastructure is deleted and
+// ready to be garbage collected by the API Server.
+MachinePhaseDeleted = MachinePhaseType("deleted")
+```
+
+#### Transition Conditions
+- `Machine.Status.Phase` is set to “Deleting”
+- `Machine.ObjectMeta.DeletionTimestamp` is not `<nil>`
+- `Machine.Spec.InfrastructureRef`->`Status.Phase` is “Deleted”
+
+#### Expectations
+- Machine controller should remove finalizer.
+
+---
+### Failed
+```go
+// MachinePhaseFailed is the Machine state when the system
+// might require user intervention.
+MachinePhaseFailed = MachinePhaseType("failed")
+```
+
+#### Transition Conditions
+- `Machine.ErrorReason` and/or `Machine.ErrorMessage` is populated.
+
+#### Expectations
+- User intervention.
+
+
+![Figure 1](https://i.imgur.com/VApZ8vv.png)
+
+
+---
+### Sequence diagram: User creates a machine with Kubeadm bootstrapper.
+In this scenario, we go through each step from “kubectl apply” to seeing the Node in “Running” state. We assume the user has chosen to create a Machine without custom user data, a Kubeadm bootstrap provider in Machine.Bootstrap, and an AWS infrastructure provider in Machine.InfrastructureRef.
+
+![Figure 2](https://i.imgur.com/1UMKnkH.png)
+
+Figure 2 is a simplified diagram of the configuration of a Machine using a Kubeadm bootstrap provider. Machine.Spec.Bootstrap.ConfigRef points to a bootstrap provider-specific custom resource; in this example, a KubeadmBootstrapConfig called controlplane-0.
+
+![Figure 3](https://i.imgur.com/dKUJoSf.png)
```

Comment:
In figure 3, the first three "opt" boxes that overlap the blue column all have the same condition ("[Required only if the object hasn't been seen before]"). I think all three of them could be a single box, since if one of them happens, then all of them happen.

Similarly, the two opt boxes that overlap both columns share a single condition, and could also be combined into a single box.